add --dry-run flag to `kubectl apply, kubectl create`

juanvallejo opened the pull request:
Partially addresses https://github.com/kubernetes/kubernetes/issues/11488

**Release note**:

``` release-note
release-note-none
```

Related StackOverflow:
http://stackoverflow.com/questions/38824409/validate-openshift-objects-defined-in-yaml-before-actually-applying-or-executing

This patch adds a `--dry-run` flag to the `apply` command in order to
allow validation of objects, without patching or creating them. If a
`--dry-run` flag is present and no validation errors are found, the
command will exit before "configuring" or creating any objects.

@ncdc @fabianofranz

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34028)

<!-- Reviewable:end -->

ncdc commented:
I'd recommend that you read through the history of #11488 to see if this meshes with the suggestions there.

juanvallejo commented:
@ncdc 
> I'd recommend that you read through the history of #11488 to see if this meshes with the suggestions there.

Thanks! It partially addresses some comments from that thread where `--validate=true --dry-run` is wanted

soltysh commented:
cc @kubernetes/kubectl

AdoHe commented:
I will comment more when I back to office.

soltysh commented:
The change itself looks good, but my expectations from using `--dry-run` are that the command works  almost identical to regular (non-dry-runned) in the output. This means you should create a visitor and call `PrintSuccess` with `dryRun` flag set to `true`. Expected output:
```
$ kubectl apply -f --dry-run pod.yaml
pod "nginx" created (dry run)
```
vs
```
$ kubectl apply -f pod.yaml
pod "nginx" created
```

Additionally, since you're at it can you also add this to `kubectl create` it looks quite similar.

juanvallejo commented:
@soltysh 
> The change itself looks good, but my expectations from using --dry-run are that the command works almost identical to regular (non-dry-runned) in the output. This means you should create a visitor and call PrintSuccess with dryRun flag set to true

Sounds good, will do

> Additionally, since you're at it can you also add this to kubectl create it looks quite similar.

Definitely, [currently working on this downstream](https://github.com/openshift/origin/pull/11238), I'll open it upstream once I have a bit more progress on it

AdoHe commented:
since we have `dry-run` on several commands, I am considering whether we should add a normal proposal for this.

juanvallejo committed:
add --dry-run flag to `kubectl apply`

Related StackOverflow:
http://stackoverflow.com/questions/38824409/validate-openshift-objects-defined-in-yaml-before-actually-applying-or-executing

This patch adds a `--dry-run` flag to the `apply` command in order to
allow validation of objects, without actually creating them. If a
`--dry-run` flag is present and no validation errors are found, the
command will exit before patching or creating any objects.

It also adds a `--dry-run` option to the `kubectl create` root command.

juanvallejo commented:
@soltysh or @AdoHe PTAL. Went ahead and updated `apply`'s `--dry-run` option to behave like other `--dry-run` options and added it to `kubectl complete as well`

AdoHe commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -94,14 +94,15 @@ func NewCmdApply(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	usage := "that contains the configuration to apply"
 	cmdutil.AddFilenameOptionFlags(cmd, &options.FilenameOptions, usage)
 	cmd.MarkFlagRequired("filename")
+	cmd.Flags().Bool("dry-run", false, "If true, only print validation for objects that would be created.")
```
since we will add this flag in multiple commands, please create a helper method like `AddDryRunFlag` in `cmdutil`

soltysh replied:
:+1:

AdoHe commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -195,47 +198,52 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *Ap
+
 			count++
-			cmdutil.PrintSuccess(mapper, shortOutput, out, info.Mapping.Resource, info.Name, false, "created")
+			cmdutil.PrintSuccess(mapper, shortOutput, out, info.Mapping.Resource, info.Name, dryRun, "created")
```
what's output of this? The resource is not actually created.

soltysh replied:
The use of this function is my suggestion. Generally, you'll get a similar output to running that command without `--dry-run` flag, iow. it'll print you `resource X created` w/wo `(dry run)` at the end.

AdoHe commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -195,47 +198,52 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *Ap
 		}
 		count++
-		cmdutil.PrintSuccess(mapper, shortOutput, out, info.Mapping.Resource, info.Name, false, "configured")
+		cmdutil.PrintSuccess(mapper, shortOutput, out, info.Mapping.Resource, info.Name, dryRun, "configured")
```
the same.

soltysh reviewed (changes requested):
Generally looks good, one nit from @AdoHe I'd like to see addressed before letting this in.

juanvallejo commented:
@AdoHe @soltysh Thanks for the review! Turns out there was already a helper function for adding a DryRun flag. Went ahead and updated this PR to use it.

soltysh commented:
LGTM, can you please squash your changes and I'll tag it for merge.

juanvallejo commented:
@soltysh 
> can you please squash your changes and I'll tag it for merge.

Done!

soltysh commented:
@k8s-bot ok to test

ymqytw commented:
@juanvallejo It seems that `kubectl create` won't respect flag `--output` when using `--dry-run`.
```
$ kubectl create -f ../pod.yaml --dry-run -o yaml
pod "test-pod" created (dry run)
```
Can you open an issue or a followup PR for it?

ymqytw commented:
@juanvallejo `kubectl apply` doesn't work well.
```
$ kubectl apply -f ../pod.yaml --dry-run -o yaml
error: Unexpected -o output mode: yaml. We only support '-o name'.
```

I think we should make `create` and `apply` work similar as `kubectl create configmap --dry-run -o yaml`.

juanvallejo commented:
@ymqytw
> It seems that kubectl create won't respect flag --output when using --dry-run.

Thanks, I'll go ahead and open an issue and followup PR for this

> kubectl apply doesn't work well.

I'll update `kubectl apply` to support more than just `-o name` in the same PR

ymqytw commented:
@juanvallejo Thanks. Hopefully we can fix it before 1.5 get released.
